cpu/nrf51/radio/nrf51_ble/nrf51_ble.c

```diff
@@ -0,0 +1,630 @@
+    NRF_RADIO->RXADDRESSES = 0x01UL;    /* listen to addresses 0 and 1 */
+
+    /* configure data fields and packet length whitening and endianess */
+    NRF_RADIO->PCNF0 = (CONF_S1 << RADIO_PCNF0_S1LEN_Pos) |
+                       (CONF_S0 << RADIO_PCNF0_S0LEN_Pos) |
+                       (CONF_LEN << RADIO_PCNF0_LFLEN_Pos);
+
+    NRF_RADIO->PCNF1 = (CONF_WHITENING << RADIO_PCNF1_WHITEEN_Pos) |
+                       (CONF_ENDIAN << RADIO_PCNF1_ENDIAN_Pos) |
+                       (CONF_BASE_ADDR_LEN << RADIO_PCNF1_BALEN_Pos) |
+                       //(CONF_STATLEN << RADIO_PCNF1_STATLEN_Pos) |
+                       (CONF_PAYLOAD_LEN << RADIO_PCNF1_MAXLEN_Pos);
+    /* configure CRC unit */
+    NRF_RADIO->CRCCNF = RADIO_CRCCNF_LEN_Three << RADIO_CRCCNF_LEN_Pos |
+                        RADIO_CRCCNF_SKIPADDR_Skip << RADIO_CRCCNF_SKIPADDR_Pos;
+
+    NRF_RADIO->CRCPOLY = CONF_CRC_POLY;
+    NRF_RADIO->CRCINIT = CONF_CRC_INIT;
+    NRF_RADIO->DATAWHITEIV = BLEMIN_DEFAULT_CHANNEL;
+
+    /* set shortcuts for more efficient transfer */
+    NRF_RADIO->SHORTS = (1 << RADIO_SHORTS_READY_START_Pos);
+    /* enable interrupts */
+    NVIC_SetPriority(RADIO_IRQn, RADIO_IRQ_PRIO);
+    NVIC_EnableIRQ(RADIO_IRQn);
+    /* enable END interrupt */
+    NRF_RADIO->EVENTS_END = 0;
+    NRF_RADIO->INTENSET = (1 << RADIO_INTENSET_END_Pos);
+
+    /* put device in receive mode */
+    _switch_to_rx();
+
+    return 0;
+}
+
+/* BLE interface low-level receive function */
+static void _receive_data(void)
+{
+    // DEBUG("nrf51_ble: receive_data\n");
+
+    ble_pkt_t *data;
+    ng_pktsnip_t *pkt;
+
+    /* only read data if we have somewhere to send it to */
+    if (_netdev->event_cb == NULL) {
+        DEBUG("nrf51_ble: receive_data (return)\n");
+        return;
+    }
+
+    /* get pointer to RX data buffer */
+    data = &(_rx_buf[_rx_next ^ 1]);
+
+    /* allocate and fill payload */
+    pkt = ng_pktbuf_add(NULL, data->payload, data->header.length, NG_NETTYPE_UNDEF);
+    pkt = ng_pktbuf_add(pkt, data, BLE_PDU_HDR_LEN, NG_NETTYPE_UNDEF);
+
+    if (pkt == NULL) {
+        DEBUG("nrf51_ble: Error allocating packet payload on RX\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    /* pass on the received packet */
+    _netdev->event_cb(NETDEV_EVENT_RX_COMPLETE, pkt);
+}
+
+/* BLE interface low-level send function */
+int _send(ng_netdev_t *dev, ng_pktsnip_t *pkt)
+{
+    // DEBUG("nrf51_ble: send\n");
+
+    (void)dev;
+    size_t size;
+
+    uint8_t payload[CONF_PAYLOAD_LEN]; /**< actual payload */
+
+    /* check packet */
+    if (pkt == NULL) {
+        DEBUG("nrf51_ble: Error sending packet: packet incomplete\n");
+        return -ENOMSG;
+    }
+
+    /* check if payload is withing length bounds */
+    size = ng_pkt_len(pkt);
+
+    if (size > CONF_PAYLOAD_LEN) {
+        ng_pktbuf_release(pkt);
+        DEBUG("nrf51_ble: Error sending packet: invalid BLE pdu header length\n");
```

Comment:
The error message indicates that the `size` is greater that `BLE_PDU_HDR_LEN` but you check for `CONF_PAYLOAD_LEN`. Is everything correct here?